sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileFormat.scala

```diff
@@ -252,6 +253,13 @@ private[orc] class OrcOutputWriter(
   override def close(): Unit = {
     if (recordWriterInstantiated) {
       recordWriter.close(Reporter.NULL)
+    } else {
+      // SPARK-15474 Write empty orc file with correct schema
+      val conf = context.getConfiguration()
+      val writer = org.apache.orc.OrcFile.createWriter(
+        new Path(path), org.apache.orc.mapred.OrcOutputFormat.buildOptions(conf))
+      new org.apache.orc.mapreduce.OrcMapreduceRecordWriter(writer)
+      writer.close()
```

Comment:
So, if i understood correctly it will write out by `org.apache.orc.mapreduce.OrcMapreduceRecordWriter` when output is empty but, write out by `org.apache.hadoop.hive.ql.io.orc.OrcRecordWriter` when output is non-empty? I thought we should use the same writer for both paths if possible and this one looks rather a band-aid fix. It won't block this PR but I wonder if this is the only way we could do for now.

Reply:
Yep. That's correct understanding. This PR intentionally focuses only on handling empty files and inferring schema. This will help us transit safely from old Hive ORC to new Apache ORC 1.4.1.